In [5]:
import openpyxl

wb = openpyxl.load_workbook('Documents/AI Engineer Learning/data/online_retail_II.xlsx', read_only=True)
ws = wb.active  # get the first sheet

print("Sheet name:", ws.title)
print("Total rows:", ws.max_row)
print("Total cols:", ws.max_column)

# Print the header row (row 1)
headers = [cell.value for cell in next(ws.iter_rows(min_row=1, max_row=1))]
print("\nColumn names:", headers)

wb.close()

Sheet name: Year 2010-2011
Total rows: 541911
Total cols: 8

Column names: ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


Day 4: A Python dict is a VLOOKUP table. The GROUP BY equivalent is building a dict inside a loop. The retail dataset has 541911 rows and columns: 'Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country'.

In [20]:
class SalesReport:
    """
    A class that holds sales data and can compute KPIs and classify performance.
    Think of this as a reusable 'template' for processing any sales dataset.
    """
    def __init__(self, transactions):
        self.transactions = transactions

    def total_revenue(self):
        return sum(t["amount"] for t in self.transactions)

    def average_order(self):
        if len(self.transactions) == 0:
            return 0
        return self.total_revenue() / len(self.transactions)

    def top_countries(self, n=3):
        country_totals = {}
        for t in self.transactions:
            c = t["country"]
            country_totals[c] = country_totals.get(c, 0) + t["amount"]
        sorted_countries = sorted(country_totals.items(),
                                   key=lambda x: x[1], reverse=True)
        return sorted_countries[:n]

    def summary(self):
        print("=== Sales Report ===")
        print(f"Total Revenue:   ${self.total_revenue():,.2f}")
        print(f"Average Order:   ${self.average_order():,.2f}")
        print(f"Total Orders:    {len(self.transactions)}")
        print(f"Top 3 Countries: {self.top_countries()}")

    def save_report(self, filename="report.txt"):
        """Saves the summary report to a text file."""
        with open(filename, "w") as f:
            f.write("=== Sales Report ===\n")
            f.write(f"Total Revenue:  ${self.total_revenue():,.2f}\n")
            f.write(f"Average Order:  ${self.average_order():,.2f}\n")
            f.write(f"Total Orders:   {len(self.transactions)}\n")
            for country, total in self.top_countries():
                f.write(f"{country}: ${total:,.2f}\n")  # ← indented inside the loop
        print(f"Report saved to {filename}")

In [21]:
sample_data = [
    {"invoice": "001", "country": "UK",      "amount": 450},
    {"invoice": "002", "country": "Germany", "amount": 820},
    {"invoice": "003", "country": "UK",      "amount": 215},
    {"invoice": "004", "country": "France",  "amount": 630},
]

# This creates ONE object (an 'instance') from the SalesReport class
report = SalesReport(sample_data)

# Now call the methods on that object
report.summary()
print("\nTop 2 Countries:", report.top_countries(n=2))

=== Sales Report ===
Total Revenue:   $2,115.00
Average Order:   $528.75
Total Orders:    4
Top 3 Countries: [('Germany', 820), ('UK', 665), ('France', 630)]

Top 2 Countries: [('Germany', 820), ('UK', 665)]


In [8]:
import requests

In [9]:
print(requests.__version__)

2.32.5


In [22]:
report.save_report("my_first_report.txt")

Report saved to my_first_report.txt


Journal: "Day 5: A class bundles data + functions. When sklearn says XGBClassifier(), it's the same pattern as SalesReport(). __init__ sets up the object. Methods are functions that belong to the class."